test/hotspot/jtreg/runtime/Monitor/TestRecursiveLocking.java

```diff
@@ -0,0 +1,560 @@
+            Asserts.assertEquals(INNER.counter, constLockStackCapacity);
+        }
+
+        synchronized void runA(int depth, SynchronizedObject B) {
+            counter++;
+
+            if (flagLockingMode == LM_LIGHTWEIGHT) {
+                // First time we lock A, A is the only one on the lock
+                // stack.
+                if (counter == 1) {
+                    assertNotInflated();
+                } else {
+                    // Second time we want to lock A, the lock stack
+                    // looks like this [A, B].  Lightweight locking
+                    // doesn't allow interleaving ([A, B, A]), instead
+                    // it inflates A and removed it from the lock
+                    // stack. Which leaves us with only [B] on the lock
+                    // stack. After more recursions it will grow to
+                    // [B, B ... B].
+                    assertInflated();
+                }
+            } else if (flagLockingMode == LM_MONITOR) {
+                assertInflated();
+            }
+
+            // Call runB() at the same depth as runA's depth:
+            B.runB(depth, this);
+        }
+
+        synchronized void runB(int depth, SynchronizedObject A) {
+            counter++;
+
+            if (flagLockingMode != LM_MONITOR) {
+                // Legacy tolerates endless recursions. While testing
+                // lightweight we don't go deeper than the size of the
+                // lock stack, which in this test case will be filled
+                // with a number of B-elements. See comment in runA()
+                // above for more info.
+                assertNotInflated();
+            } else {
+                assertInflated();
+            }
+
+            if (depth == 1) {
+                // Reached LockStackCapacity in depth so we're done.
+                return;
+            } else {
+                A.runA(depth - 1, this);
+            }
+        }
+
+        // This test alternates by locking A and B.
+        public void runAlternateABTest() {
+            final SynchronizedObject A = new SynchronizedObject();
+            final SynchronizedObject B = new SynchronizedObject();
+
+            // Just checking since they are new objects:
+            A.assertNotInflated();
+            B.assertNotInflated();
+
+            A.runA(constLockStackCapacity, B);
+
+            // Verify that the nested monitors have been properly released:
+            syncThread.verifyCanBeSynced(A);
+            syncThread.verifyCanBeSynced(B);
+
+            Asserts.assertEquals(A.counter, constLockStackCapacity);
+            Asserts.assertEquals(B.counter, constLockStackCapacity);
+            if (flagLockingMode == LM_LEGACY) {
+                A.assertNotInflated();
+            }
+            // Implied else: for LM_MONITOR or LM_LIGHTWEIGHT it can be
+            // either inflated or not point because A is not locked anymore
```

Comment:
nit typo: s/either inflated or not point because/either inflated or not because/

Reply:
fixed